finite_difference/src/subdomain_mod.f90

```diff
@@ -0,0 +1,302 @@
+       end if
+       hwidth = halo_width
+    end if
+
+    decomp%ndomains = ndom
+    allocate(decomp%subdomains(ndom), Stat=ierr)
+    if(ierr /= 0)then
+       call parallel_abort('decompose: failed to allocate tiles array')
+    end if
+
+    xlen = domainx
+    ylen = domainy
+    
+    if(auto_tile)then
+
+       ntilex = INT( SQRT(REAL(ndom)) )
+       do WHILE(MOD(ndom, ntilex) /= 0)
+          ntilex = ntilex - 1
+       end do
+       ntiley = ndom/ntilex
+
+       ! Match longest dimension of domain to longest dimension of 
+       ! process grid
+       if(xlen > ylen)then
+          if( ntilex < ntiley )then
+             ierr   = ntiley
+             ntiley = ntilex
+             ntilex = ierr
+          end if
+       else
+          ! N >= M so want nprocy >= nprocx
+          if( ntiley < ntilex )then
+             ierr   = ntiley
+             ntiley = ntilex
+             ntilex = ierr
+          end if
+       end if
+    else
+       ntilex = ndomainx
+       ntiley = ndomainy
+    end if ! automatic determination of tiling grid
+
+    WRITE (*,"('decompose: using grid of ',I3,'x',I3)") ntilex, ntiley
+    decomp%nx = ntilex
+    decomp%ny = ntiley
+
+    ! If we think about the internal regions of the tiles,
+    ! then they should share the domain between them:
+    internal_width = NINT(REAL(xlen) / REAL(ntilex))
+    internal_height = NINT(REAL(ylen) / REAL(ntiley))
+
+    ! Integer arithmetic means that ntiley tiles of height idy might
+    ! actually span a height greater or less than N. If so, we try and
+    ! reduce the height of each row by just one until we've accounted
+    ! for the <jover> extra rows.
+    !nwidth = (ntiley-2)*(idy-2) + 2*(idy-1)
+    nwidth = ntiley * internal_height
+    if(nwidth > ylen)then
+       jover  = nwidth - ylen
+       junder = 0
+    else if(nwidth < ylen)then
+       jover  = 0
+       junder = ylen - nwidth
+    else
+       jover  = 0
+       junder = 0
+    end if
+    ! Ditto for x dimension
+    !nwidth = (ntilex-2)*(idx-2) + 2*(idx-1)
+    nwidth = ntilex * internal_width
+    if(nwidth > xlen)then
+       iover  = nwidth - xlen
+       iunder = 0
+    else if(nwidth < xlen)then
+       iover  = 0
+       iunder = xlen - nwidth
+    else
+       iover  = 0
+       iunder = 0
+    end if
+
+    if(print_tiles)then
+       WRITE(*,"('Tile width = ',I4,', tile height = ',I4)") &
+            internal_width, internal_height
+       WRITE(*,"('iover = ',I3,', iunder = ',I3)") iover, iunder
+       WRITE(*,"('jover = ',I3,', junder = ',I3)") jover, junder
+    end if
+
+    ith = 1
+    ! The starting point of the tiles in y
+    jval = 1
+
+    nvects_max = 0
+    nvects_min = 1000000
+    nvects_sum = 0
+    decomp%max_width  = 0
+    decomp%max_height = 0
+
+    if(print_tiles)WRITE(*,"(/'Sub-domains:')")
+
+    do jj = 1, ntiley, 1
+
+       ! Point to subdomain we are about to define
+       subdomain => decomp%subdomains(ith)
```

Comment:
Remove - not used here, and it is being redefined later.

Reply:
True but the compiler does not know that `ntilex` is guaranteed to be > 1 (and thus that the `ji` loop will do at least one trip) and therefore warns about the assignment using the `subdomain` pointer after the loop over `ji`. This is a simple way of preventing that warning.